skstore/ts/src/skstore_api.ts

```diff
@@ -0,0 +1,588 @@
+   * @returns The results of the lookup
+   * @throws {TableIndexError} when an index is not found
+   *          or the index type is not valid
+   */
+  // TODO get(key: TJSON, index?: string): R[];
+
+  /**
+   * Map over each table entry and apply mapper function
+   * @param {EntryMapper} mapper - function to apply to the table entry
+   * @returns {EHandle} The the resulting eager reactive map handle
+   */
+  map<K extends TJSON, V extends TJSON>(
+    mapper: EntryMapper<R, K, V>,
+  ): EHandle<K, V>;
+}
+
+export interface Table<R extends TJSON[]> {
+  getName(): string;
+  /**
+   * Insert an entry in the table
+   * @param entries - the entries to insert in the table
+   * @param update - indicate if the value need to be updated in case of index conflict
+   * @throws {Error} when not update and an index contraints is broken
+   */
+  insert(entry: R[], update?: boolean): void;
+  /**
+   * Update an entry in the table
+   * @param row - the table entry to update
+   * @param updates - the column values updates
+   * @throws {Error} when an index contraints is broken
+   */
+  update(row: R, updates: JSONObject): void;
+  /**
+   * Update an entry in the table
+   * @param where - the column values to filter entries
+   * @param updates - the column values updates
+   * @throws {Error} when an index contraints is broken
+   */
+  updateWhere(where: JSONObject, updates: JSONObject): void;
+  /**
+   * Select entries in the table
+   * @param where - the column values to filter entries
+   * @throws {Error} when an index contraints is broken
+   */
+  select(where: JSONObject, columns?: string[]): JSONObject[];
+  /**
+   * Delete an entry in the table
+   * @param entry - the entry to delete from the table
+   */
+  delete(entry: R): void;
+  /**
+   * Delete entries in the table
+   * @param where - the column values to filter entries
+   */
+  deleteWhere(where: JSONObject): void;
+  /**
+   * Add a watch to the table
+   * @param update - the function to call on table update
+   */
+  watch: (update: (rows: JSONObject[]) => void) => { close: () => void };
+  /**
+   * Add a watch changes to the table
+   * @param init - the function to call on table init/reset
+   * @param update - the function to call on table update
+   */
+  watchChanges: (
+    init: (rows: JSONObject[]) => void,
+    update: (added: JSONObject[], removed: JSONObject[]) => void,
+  ) => { close: () => void };
+}
+
+export type Mapping<
+  K1 extends TJSON,
+  V1 extends TJSON,
+  K2 extends TJSON,
+  V2 extends TJSON,
+> = [EHandle<K1, V1>, Mapper<K1, V1, K2, V2>];
+
+export type Lazy<K extends TJSON, V extends TJSON> = (
+  selfHdl: LHandle<K, V>,
+  key: K,
+) => Opt<V>;
+
+export interface SKStoreFactory extends Shared {
+  runSKStore(
+    init: (skstore: SKStore, ...tables: TableHandle<TJSON[]>[]) => void,
+    app: App,
+    tables: MirrorSchema[],
+    connect?: boolean,
+  ): Promise<Table<TJSON[]>[]>;
+}
+
+export interface SKStore {
```

Comment:
I'm also a bit confused about the module/interface structure here.  Why are these specific functions (`lazy`, `multimap`, `multimapReduce`, `asyncLazy`, and `log`) members of this SKStore interface and everything else at the top-level?